controllers/provisioning_controller.go

```diff
@@ -153,39 +153,45 @@ func (r *ProvisioningReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 		return ctrl.Result{}, nil
 	}
 
-	// examine DeletionTimestamp to determine if object is under deletion
-	if baremetalConfig.ObjectMeta.DeletionTimestamp.IsZero() {
-		// The object is not being deleted, so if it does not have our finalizer,
-		// then lets add the finalizer and update the object. This is equivalent
-		// registering our finalizer.
-		if !containsString(baremetalConfig.ObjectMeta.Finalizers,
-			metal3iov1alpha1.ProvisioningFinalizer) {
-			// Add finalizer becasue it doesn't already exist
-			baremetalConfig.ObjectMeta.Finalizers = append(baremetalConfig.ObjectMeta.Finalizers,
-				metal3iov1alpha1.ProvisioningFinalizer)
-			if err := r.Client.Update(context.Background(), baremetalConfig); err != nil {
-				return ctrl.Result{}, errors.Wrap(err, "failed to update Provisioning CR with its finalizer")
-			}
-		}
-	} else {
-		// The Provisioning object is being deleted
-		if containsString(baremetalConfig.ObjectMeta.Finalizers, metal3iov1alpha1.ProvisioningFinalizer) {
-			// Add any specific deletion logic here
-
-			// Remove our finalizer from the list and update it.
-			baremetalConfig.ObjectMeta.Finalizers = removeString(baremetalConfig.ObjectMeta.Finalizers,
-				metal3iov1alpha1.ProvisioningFinalizer)
-			if err := r.Client.Update(context.Background(), baremetalConfig); err != nil {
-				return ctrl.Result{}, errors.Wrap(err, "failed to remove finalizer from Provisioning CR")
-			}
+	// Read container images from Config Map
+	var containerImages provisioning.Images
+	if err := provisioning.GetContainerImages(&containerImages, r.ImagesFilename); err != nil {
+		// Images config map is not valid
+		// Provisioning configuration is not valid.
+		// Requeue request.
+		r.Log.Error(err, "invalid contents in images Config Map")
+		co_err := r.updateCOStatus(ReasonInvalidConfiguration, err.Error(), "invalid contents in images Config Map")
+		if co_err != nil {
+			return ctrl.Result{}, fmt.Errorf("unable to put %q ClusterOperator in Degraded state: %v", clusterOperatorName, co_err)
 		}
+		return ctrl.Result{}, err
 	}
 
+	info := r.provisioningInfo(baremetalConfig, &containerImages)
+
+	// Make sure ClusterOperator exists and updated
 	err = r.ensureClusterOperator(baremetalConfig)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
+	// Check if Provisioning Configuartion is being deleted
+	deleted, err := r.checkForCRDeletion(info)
+	if err != nil {
+		err = r.updateCOStatus(ReasonDeployTimedOut, err.Error(), "Unable to delete a metal3 resource on Provisioning CR deletion")
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("unable to put %q ClusterOperator in Degraded state: %v", clusterOperatorName, err)
+		}
+		return ctrl.Result{}, err
+	}
+	if err == nil && deleted {
+		err = r.updateCOStatus(ReasonComplete, "all Metal3 resources deleted", "")
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("unable to put %q ClusterOperator in Available state: %v", clusterOperatorName, err)
```

Comment:
```suggestion
			return ctrl.Result{}, fmt.Errorf("unable to put %q ClusterOperator in Available state: %w", clusterOperatorName, err)
```

Reply:
i believe that when wrapping errors we should be using %w as this allows the error to be unwrapped (please check in the rest of the PR too)

Reply:
Done.